vulnfeeds/git/repository.go

```diff
@@ -154,6 +155,34 @@ func RepoTags(repoURL string, repoTagsCache RepoTagsCache) (tags Tags, e error)
 	return tags, nil
 }
 
+// normalizeRepoTag returns a repo tag normalized.
+// It is:
+//   - lowercased,
+//   - the repo name, if present is removed
+//   - any Java package name prefix, if present is removed
+//
+// finally, it is run through the standard version normalizing treatment
+func normalizeRepoTag(tag string, reponame string) (normalizedTag string, err error) {
+	// Match the likes of "org.apache.sling.i18n-2.0.2" as seen in github.com/apache/sling-org-apache-sling-i18n
+	var javaPackageRegex = regexp.MustCompile(`(?i)^(?:[a-z][a-z0-9_]*(?:\.[a-z0-9_]+)+[0-9a-z_])(.*)$`)
+	// Opportunistically remove parts determined to match the repo name,
+	// to ease particularly difficult to normalize cases like 'openj9-0.38.0'.
+	prenormalizedTag := strings.TrimPrefix(strings.ToLower(tag), reponame)
+	// Deal with the reponame being in the *middle* of the tag like 'hudson-yui2-2800'
+	if strings.Contains(prenormalizedTag, reponame) {
+		_, after, found := strings.Cut(prenormalizedTag, reponame)
+		if found {
+			prenormalizedTag = after
+		}
+	}
+	if javaPackageRegex.MatchString(prenormalizedTag) {
+		prenormalizedTag = javaPackageRegex.FindStringSubmatch(prenormalizedTag)[1]
```

Comment:
nit: Can we just call `FindStringSubmatch` once, and check for nil? Rather than doing hte initial `MatchString` check.

Reply:
I think that would overall be worse for conciseness, as I wouldn't be able to iteratively reuse the `prenormalizedTag` variable the way I currently am, because it could get turned to `nil` by the call to `FindStringSubmatch`.